### PR TITLE
【PIR API adaptor No.35、40】 Migrate paddle.nn.ChannelShuffle/ClipGradByNorm into pir

### PR DESCRIPTION
**Pull request by fsczz**
### PR types
Others

### PR changes
APIs

### Description

PIR API 推全升级
 `paddle.nn.ClipGradByNorm ` 引用了clip_by_norm，对clip_by_norm迁移升级至 pir，并更新单测，test_clip_by_norm_op单测覆盖率：4/4   test_gradient_clip 单测覆盖率：2/2
 `paddle.nn.ChannelShuffle` 引用了paddle.nn.functional.channel_shuffle，对channel_shuffle迁移升级至 pir，并更新单测， 单测覆盖率：6/6（TestChannelShuffleError 通过）

### CODE REVIEW

**Comment by MarioLulab** [comment]
https://github.com/PaddlePaddle/Paddle/issues/58067

问题已收到，正在排查

**Comment by MarioLulab** [comment]
`test_gradient_clip.py` 文件下的 `TestGradientClipByNorm` 单测也需要迁移。具体做法为：对 TestGradientClipByNorm 新增添一个 TestPirGradientClipByNorm 单测。可以参考该文件下 `TestPirGradientClipByGlobalNorm` 与 `TestGradientClipByGlobalNorm` 的做法。
paddle.static.nn.fc 为静态图专用 api，需要改写成 paddle.nn.Linear 和 paddle.nn.functional.relu 或 paddle.nn.functional.softmax 的形式

**Comment by fsczz** [comment]
已经添加了TestPirGradientClipByNorm 并测试通过

**Review by MarioLulab — COMMENTED** [review]
nice work~
但 ClipGradByNorm 的 pir 单测还需要适配 test_none_grad 这个单测，否则 ci 覆盖率不达标
image

@0x45f 帮忙看一下这个问题：pir.Value 没有原来的 "need_clip" 属性，导致 pir 的单测与旧ir单测无法对齐，ci 覆盖率不达标，是否需要删除 `_pir_clip` 函数中对 need_clip 属性的判断？

**Comment by 0x45f** [comment]
> nice work~ 但 ClipGradByNorm 的 pir 单测还需要适配 test_none_grad 这个单测，否则 ci 覆盖率不达标 image
> 
> @0x45f 帮忙看一下这个问题：pir.Value 没有原来的 "need_clip" 属性，导致 pir 的单测与旧ir单测无法对齐，ci 覆盖率不达标，是否需要删除 `_pir_clip` 函数中对 need_clip 属性的判断？

这里可以尝试在python/paddle/pir/core.py文件create_parameter函数最后return之前添加下面两行代码：
```python
need_clip = kwargs.get('need_clip', True)
setattr(param, 'need_clip', need_clip)
```

**Comment by fsczz** [comment]
> > nice work~ 但 ClipGradByNorm 的 pir 单测还需要适配 test_none_grad 这个单测，否则 ci 覆盖率不达标 image
> > @0x45f 帮忙看一下这个问题：pir.Value 没有原来的 "need_clip" 属性，导致 pir 的单测与旧ir单测无法对齐，ci 覆盖率不达标，是否需要删除 `_pir_clip` 函数中对 need_clip 属性的判断？
> 
> 这里可以尝试在python/paddle/pir/core.py文件create_parameter函数最后return之前添加下面两行代码：
> 
> ```python
> need_clip = kwargs.get('need_clip', True)
> setattr(param, 'need_clip', need_clip)
> ```

添加python/paddle/pir/core.py文件代码之后，还需要在TestPirGradientClipByNorm的测试中添加test_none_grad单侧么
```
 def test_none_grad(self):
        clip = paddle.nn.ClipGradByNorm(self.clip_norm)
        x = (
            base.default_main_program()
            .global_block()
            .create_parameter(
                name="x", shape=[2, 3], dtype="float32", need_clip=False
            )
        )
        y = (
            base.default_main_program()
            .global_block()
            .create_parameter(
                name="y", shape=[2, 3], dtype="float32", need_clip=False
            )
        )
        # (x, None) should not be returned
        params_grads = [(x, None), (x, y)]
        params_grads = clip(params_grads)
        self.assertTrue(
            len(clip(params_grads)) == 1,
            "ClipGradByNorm: when grad is None, it shouldn't be returned by gradient clip!",
        )
        self.assertTrue(
            params_grads[0][1].name == 'y',
            "ClipGradByNorm: grad should not be clipped when filtered out!",
        )
```

**Comment by MarioLulab** [comment]
> > > nice work~ 但 ClipGradByNorm 的 pir 单测还需要适配 test_none_grad 这个单测，否则 ci 覆盖率不达标 image
> > > @0x45f 帮忙看一下这个问题：pir.Value 没有原来的 "need_clip" 属性，导致 pir 的单测与旧ir单测无法对齐，ci 覆盖率不达标，是否需要删除 `_pir_clip` 函数中对 need_clip 属性的判断？
> > 
> > 
> > 这里可以尝试在python/paddle/pir/core.py文件create_parameter函数最后return之前添加下面两行代码：
> > ```python
> > need_clip = kwargs.get('need_clip', True)
> > setattr(param, 'need_clip', need_clip)
> > ```
> 
> 添加python/paddle/pir/core.py文件代码之后，还需要在TestPirGradientClipByNorm的测试中添加test_none_grad单侧么
> 
> ```
>  def test_none_grad(self):
>         clip = paddle.nn.ClipGradByNorm(self.clip_norm)
>         x = (
>             base.default_main_program()
>             .global_block()
>             .create_parameter(
>                 name="x", shape=[2, 3], dtype="float32", need_clip=False
>             )
>         )
>         y = (
>             base.default_main_program()
>             .global_block()
>             .create_parameter(
>                 name="y", shape=[2, 3], dtype="float32", need_clip=False
>             )
>         )
>         # (x, None) should not be returned
>         params_grads = [(x, None), (x, y)]
>         params_grads = clip(params_grads)
>         self.assertTrue(
>             len(clip(params_grads)) == 1,
>             "ClipGradByNorm: when grad is None, it shouldn't be returned by gradient clip!",
>         )
>         self.assertTrue(
>             params_grads[0][1].name == 'y',
>             "ClipGradByNorm: grad should not be clipped when filtered out!",
>         )
> ```

在修改了 core.py 后，还需要在 TestPirGradientClipByNorm 的测试中添加 test_none_grad 单测。
原有的:
```python
x = (
    base.default_main_program()
    .global_block()
    .create_parameter(
        name="x", shape=[2, 3], dtype="float32", need_clip=False
    )
)
```

可以改成:
```python
with paddle.pir_utils.IrGuard():
    main = paddle.static.Program()
    startup = paddle.static.Program()
    with paddle.static.program_guard(main, startup):
        x = paddle.pir.core.create_parameter(
            dtype="float32",
            shape=[2, 3],
            name="x",
            initializer=paddle.nn.initializer.Constant(value=0.5),
            need_clip=False
        )
```

**Comment by fsczz** [comment]
在core.py中添加代码后报错
image
param 信息：
image

注释core.py代码，添加test_none_grad
```
    def test_none_grad(self):
        clip = paddle.nn.ClipGradByNorm(self.clip_norm)
        with paddle.pir_utils.IrGuard():
            main = paddle.static.Program()
            startup = paddle.static.Program()
            with paddle.static.program_guard(main, startup):
                x = paddle.pir.core.create_parameter(
                    dtype="float32",
                    shape=[2, 3],
                    name="x",
                    initializer=paddle.nn.initializer.Constant(value=0.5),
                    need_clip=False
                )
                y = paddle.pir.core.create_parameter(
                    dtype="float32",
                    shape=[2, 3],
                    name="y",
                    initializer=paddle.nn.initializer.Constant(value=0.3),
                    need_clip=False
                )
                # (x, None) should not be returned
                params_grads = [(x, None), (x, y)]
                params_grads = clip(params_grads)
                self.assertTrue(
                    len(clip(params_grads)) == 1,
                    "ClipGradByNorm: when grad is None, it shouldn't be returned by gradient clip!",
                )

                self.assertTrue(
                    params_grads[0][1].name == 'y',
                    "ClipGradByNorm: grad should not be clipped when filtered out!",
                )
```
显示没有name信息，
image

**Comment by MarioLulab** [comment]
关于给 pybind 导出的 C++ 对象设置动态属性的方法，可以参考：https://github.com/PaddlePaddle/Paddle/pull/60345
该 pr 可以解决当前使用 ↓ 遇到的问题
```python
setattr(param, 'need_clip', need_clip)
```

可以等该 pr https://github.com/PaddlePaddle/Paddle/pull/60345 合入后再推进此 pr

File: python/paddle/nn/clip.py [hunk 1]
```diff
@@ -540,6 +540,18 @@ def _dygraph_clip(self, params_grads):
             params_and_grads.append((p, new_grad))
         return params_and_grads
 
+    def _pir_clip(self, params_grads):
+        params_and_grads = []
+        for p, g in params_grads:
+            if g is None:
+                continue
+            if getattr(p, 'stop_gradient', True) is True:
```
Comment (MarioLulab) [inline python/paddle/nn/clip.py]:
```suggestion
            if getattr(p, 'need_clip', True) is False:
```
这行要保持和旧ir静态图的行为一致，都是去检查 `need_clip` 属性

**Review by MarioLulab — COMMENTED** [review]
辛苦 merge 一下最新分支，然后可以尝试在python/paddle/pir/core.py文件create_parameter函数参考 https://github.com/PaddlePaddle/Paddle/pull/60345 对 regulizer 属性的添加方法，添加下面两行代码：
```python
need_clip = kwargs.get('need_clip', True)
param.need_clip = need_clip
```

**Comment by MarioLulab** [comment]
想问下大佬的这个 pr 为啥关掉了 🐶  还在继续推进吗
